coins: don't mutate main cache when connecting block

andrewtoth opened the pull request:
This is a slightly modified version of the first few commits of #31132, which can be merged as an independent change. It has a small benefit on its own, but will help in moving the parent PR forward.

When accessing coins via the `CCoinsViewCache`, methods like `GetCoin` can call `FetchCoin` which actually mutate `cacheCoins` internally to cache entries when they are pulled from the backing db. This is generally a performance improvement for single threaded access patterns, but it precludes us from accessing entries in a `CCoinsViewCache` from multiple threads without a lock.

Another aspect is that when we use the resettable `CCoinsViewCache` view backed by the main cache for use in `ConnectBlock()`, we will insert entries into the main cache even if the block is determined to be invalid. This is not the biggest concern, since an invalid block requires proof-of-work. But, an attacker could craft multiple invalid blocks to fill the main cache. This would make us `Flush` the cache more often than necessary. Obviously this would be very expensive to do on mainnet.

Introduce `CoinsViewOverlay`, a `CCoinsViewCache` subclass that reads coins without mutating the underlying cache via `FetchCoin()`.

Add `PeekCoin()` to look up a Coin through a stack of `CCoinsViewCache` layers without populating parent caches. This prevents the main cache from caching inputs pulled from disk for a block that has not yet been fully validated. Once `Flush()` is called on the view, these inputs will be added as spent to `coinsCache` in the main cache via `BatchWrite()`.

This is the foundation for async input fetching, where worker threads must not mutate shared state.

andrewtoth committed:
coins: add PeekCoin()

Introduce a helper to look up a Coin through a stack of CCoinsViewCache layers without populating parent caches.

This is useful for ephemeral views (e.g. during ConnectBlock) that want to avoid polluting CoinsTip() when validating invalid blocks.

Co-authored-by: l0rinc <pap.lorinc@gmail.com>
Co-authored-by: Pieter Wuille <pieter@wuille.net>
Co-authored-by: Ryan Ofsky <ryan@ofsky.org>

andrewtoth committed:
fuzz: pass coins_view_cache to TestCoinsView in coins_view

Refactor TestCoinsView() to accept the cache as a parameter instead of
creating it internally. This prepares for adding a CoinsViewOverlay
fuzz target that needs to pass in a different cache type.

This is a non-functional change.

Co-authored-by: l0rinc <pap.lorinc@gmail.com>

l0rinc reviewed (changes requested):
The implementation talks about block connection but the implementation was added to block disconnect. I don't yet know if we should apply it there or not, but I find it very likely that we should do it for block connection.
A lot of tests were added for testing the new cache in isolation and they all passed while it was applied to the wrong place as far as I can tell.

We should add a new integration test that actually checks that invalid blocks don't pollute the underlying cache:
```patch
diff --git a/src/test/validation_chainstate_tests.cpp b/src/test/validation_chainstate_tests.cpp
--- a/src/test/validation_chainstate_tests.cpp	(revision 20f0b8304d089d641060f693d3284e69524fbcbe)
+++ b/src/test/validation_chainstate_tests.cpp	(date 1767360120004)
@@ -3,10 +3,12 @@
 // file COPYING or http://www.opensource.org/licenses/mit-license.php.
 //
 #include <chainparams.h>
+#include <consensus/amount.h>
 #include <consensus/validation.h>
 #include <node/kernel_notifications.h>
 #include <random.h>
 #include <rpc/blockchain.h>
+#include <script/script.h>
 #include <sync.h>
 #include <test/util/chainstate.h>
 #include <test/util/coins.h>
@@ -63,6 +65,30 @@
     }
 }
 
+BOOST_FIXTURE_TEST_CASE(connect_tip_does_not_cache_inputs_on_failed_connect, TestChain100Setup)
+{
+    Chainstate& chainstate = Assert(m_node.chainman)->ActiveChainstate();
+
+    COutPoint outpoint;
+    {
+        LOCK(cs_main);
+        outpoint = AddTestCoin(m_rng, chainstate.CoinsTip());
+        chainstate.CoinsTip().Flush(/*will_reuse_cache=*/false);
+    }
+
+    CMutableTransaction tx;
+    tx.vin.emplace_back(outpoint);
+    tx.vout.emplace_back(MAX_MONEY, CScript{} << OP_TRUE);
+
+    const auto tip{WITH_LOCK(cs_main, return chainstate.m_chain.Tip()->GetBlockHash())};
+    CBlock block = CreateBlock({tx}, CScript{} << OP_TRUE, chainstate);
+    Assert(m_node.chainman)->ProcessNewBlock(std::make_shared<CBlock>(block), true, true, nullptr);
+
+    LOCK(cs_main);
+    BOOST_CHECK_EQUAL(tip, chainstate.m_chain.Tip()->GetBlockHash()); // block rejected
+    BOOST_CHECK(!chainstate.CoinsTip().HaveCoinInCache(outpoint));    // input not cached
+}
+
 //! Test UpdateTip behavior for both active and background chainstates.
 //!
 //! When run on the background chainstate, UpdateTip should do a subset
diff --git a/src/validation.cpp b/src/validation.cpp
--- a/src/validation.cpp	(revision 20f0b8304d089d641060f693d3284e69524fbcbe)
+++ b/src/validation.cpp	(date 1767360154724)
@@ -3102,7 +3102,7 @@
     LogDebug(BCLog::BENCH, "  - Load block from disk: %.2fms\n",
              Ticks<MillisecondsDouble>(time_2 - time_1));
     {
-        CCoinsViewCache view(&CoinsTip());
+        CoinsViewCacheNonMutating view(&CoinsTip());
         bool rv = ConnectBlock(*block_to_connect, state, pindexNew, view);
         if (m_chainman.m_options.signals) {
             m_chainman.m_options.signals->BlockChecked(block_to_connect, state);
```

Concept ACK (edited to remove the nack).

l0rinc commented on src/test/coinsviewcachenonmutating_tests.cpp:
```diff
@@ -0,0 +1,143 @@
+        CMutableTransaction tx;
+        Txid txid{Txid::FromUint256(uint256(i))};
+        tx.vin.emplace_back(txid, 0);
+        prevhash = tx.GetHash();
```
This seems like a leftover.

l0rinc reviewed (approved):
I'll run a reindex-chainstate on these to see how they behave - as long as they don't slow down anything, I would be fine with them.

l0rinc commented on src/coinsviewcachenonmutating.h:
```diff
@@ -0,0 +1,40 @@
+    CCoinsMap::iterator FetchCoin(const COutPoint& outpoint) const override
+    {
+        auto [ret, inserted]{cacheCoins.try_emplace(outpoint)};
+        if (!inserted) return ret;
```
Hmm, we're we're duplicating the `FetchCoin` implementation here - and also making it virtual which introduces a slight regression on the hot path.
Could we extract only the part that changes?
I think we can simplify it by adding a virtual `GetCoinFromBase` which either calls `base->GetCoin(outpoint)` (default impl) or in `CoinsViewCacheNonMutating` it could be overwritten with the new `base->PeekCoin(outpoint)`. That way only the `if (inserted)` branch would pay for virtual dispatch and there would be no duplicated `FetchCoin` logic, something like:
```patch
diff --git a/src/coins.cpp b/src/coins.cpp
--- a/src/coins.cpp	(revision 85b20448640f8209e453019c3e1494c1ae229d74)
+++ b/src/coins.cpp	(revision c3b34ec796a3fd6e34e0dfb862c61bc298ba8d6f)
@@ -59,10 +59,15 @@
     return memusage::DynamicUsage(cacheCoins) + cachedCoinsUsage;
 }
 
+std::optional<Coin> CCoinsViewCache::GetCoinFromBase(const COutPoint& outpoint) const
+{
+    return base->GetCoin(outpoint);
+}
+
 CCoinsMap::iterator CCoinsViewCache::FetchCoin(const COutPoint &outpoint) const {
     const auto [ret, inserted] = cacheCoins.try_emplace(outpoint);
     if (inserted) {
-        if (auto coin{base->GetCoin(outpoint)}) {
+        if (auto coin{GetCoinFromBase(outpoint)}) {
             ret->second.coin = std::move(*coin);
             cachedCoinsUsage += ret->second.coin.DynamicMemoryUsage();
             if (ret->second.coin.IsSpent()) { // TODO GetCoin cannot return spent coins
diff --git a/src/coins.h b/src/coins.h
--- a/src/coins.h	(revision 85b20448640f8209e453019c3e1494c1ae229d74)
+++ b/src/coins.h	(revision c3b34ec796a3fd6e34e0dfb862c61bc298ba8d6f)
@@ -382,6 +382,8 @@
     /* Cached dynamic memory usage for the inner Coin objects. */
     mutable size_t cachedCoinsUsage{0};
 
+    virtual std::optional<Coin> GetCoinFromBase(const COutPoint& outpoint) const;
+
 public:
     CCoinsViewCache(CCoinsView *baseIn, bool deterministic = false);
 
@@ -489,7 +491,7 @@
      * @note this is marked const, but may actually append to `cacheCoins`, increasing
      * memory usage.
      */
-    virtual CCoinsMap::iterator FetchCoin(const COutPoint &outpoint) const;
+    CCoinsMap::iterator FetchCoin(const COutPoint &outpoint) const;
 };
 
 //! Utility function to add all of a transaction's outputs to a cache.
diff --git a/src/coinsviewcachenonmutating.h b/src/coinsviewcachenonmutating.h
--- a/src/coinsviewcachenonmutating.h	(revision 85b20448640f8209e453019c3e1494c1ae229d74)
+++ b/src/coinsviewcachenonmutating.h	(revision c3b34ec796a3fd6e34e0dfb862c61bc298ba8d6f)
@@ -9,28 +9,16 @@
 #include <primitives/transaction.h>
 
 /**
- * CCoinsViewCache subclass that does not call GetCoin on base via FetchCoin, so base will not be mutated before Flush.
+ * CCoinsViewCache subclass that does not call GetCoin on base, so base will not be mutated before Flush.
  * Only used in ConnectBlock to pass as an ephemeral view that can be reset if the block is invalid.
- * It provides the same interface as CCoinsViewCache, overriding the FetchCoin private method.
+ * It provides the same interface as CCoinsViewCache, overriding GetCoinFromBase().
  */
 class CoinsViewCacheNonMutating : public CCoinsViewCache
 {
 private:
-    CCoinsMap::iterator FetchCoin(const COutPoint& outpoint) const override
+    std::optional<Coin> GetCoinFromBase(const COutPoint& outpoint) const override
     {
-        auto [ret, inserted]{cacheCoins.try_emplace(outpoint)};
-        if (!inserted) return ret;
-
-        // TODO: Remove spent checks once we no longer return spent coins in coinscache_sim CoinsViewBottom.
-        if (auto coin{base->PeekCoin(outpoint)}; coin && !coin->IsSpent()) {
-            ret->second.coin = std::move(*coin);
-        } else {
-            cacheCoins.erase(ret);
-            return cacheCoins.end();
-        }
-
-        cachedCoinsUsage += ret->second.coin.DynamicMemoryUsage();
-        return ret;
+        return base->PeekCoin(outpoint);
     }
 
 public:
```

l0rinc commented on src/validation.cpp:
```diff
@@ -3098,7 +3099,7 @@ bool Chainstate::ConnectTip(
              Ticks<MillisecondsDouble>(time_2 - time_1));
     {
-        CCoinsViewCache view(&CoinsTip());
+        CoinsViewCacheNonMutating view(&CoinsTip());
```
here or in a follow-up we could add this to `TestBlockValidity` as well, where it would be even more relevant since it’s invoked with without PoW checks (so your PR arguments apply here even more)

l0rinc commented on src/test/coinsviewcachenonmutating_tests.cpp:
```diff
@@ -0,0 +1,140 @@
+            BOOST_CHECK(!main_cache.HaveCoinInCache(in.prevout));
+        }
+    }
+}
```
Checked the coverage of this and compared it with the suggested `CCoinsViewCache::GetCoinFromBase` - existing `FetchCoin` is covered that way.
We could extend the coverage with `HaveCoin()`/`GetCoin()` as well, to check that they also do not populate the parent cache when reading inputs from disk, and that flushing an ephemeral `CoinsViewCacheNonMutating` layer propagates spentness to its parent:
```patch
diff --git a/src/test/coinsviewcachenonmutating_tests.cpp b/src/test/coinsviewcachenonmutating_tests.cpp
--- a/src/test/coinsviewcachenonmutating_tests.cpp	(revision c3b34ec796a3fd6e34e0dfb862c61bc298ba8d6f)
+++ b/src/test/coinsviewcachenonmutating_tests.cpp	(revision 3e8676d49ecf9e3c12cd641ce655e74f85f3a6e6)
@@ -82,6 +82,13 @@
     PopulateView(block, db);
     CCoinsViewCache main_cache{&db};
     CoinsViewCacheNonMutating view{&main_cache};
+    const auto& outpoint{block.vtx[1]->vin[0].prevout};
+
+    BOOST_CHECK(view.HaveCoin(outpoint));
+    BOOST_CHECK(view.GetCoin(outpoint).has_value());
+    BOOST_CHECK(!main_cache.HaveCoinInCache(outpoint));
+
     CheckCache(block, view);
     // Check that no coins have been moved up to main cache from db
     for (const auto& tx : block.vtx) {
@@ -89,6 +96,12 @@
             BOOST_CHECK(!main_cache.HaveCoinInCache(in.prevout));
         }
     }
+
+    view.SetBestBlock(uint256::ONE);
+    BOOST_CHECK(view.SpendCoin(outpoint));
+    view.Flush();
+    BOOST_CHECK(!main_cache.PeekCoin(outpoint).has_value());
 }
 
 BOOST_AUTO_TEST_CASE(fetch_inputs_from_cache)
```

willcl-ark reviewed (commented):
Nice carve-out from #31132.

Left a comment about whether we want to add a CoinsViewErrorCatcher wrapper for PeekCoin, but otherwise looks good to me.

l0rinc reviewed (changes requested):
Thanks for fixing it so quickly. 
I think we should remove the `noexcept` from `GetCoinFromBase` as well because the underlying `PeekCoin` -> `GetCoin` path can still throw.
And since `PeekCoin` **does** actually call `GetCoin`, I think we need to update the PR description and comments and commit messages to accurately reflect that we're using `PeekCoin` to avoid `FetchCoin`'s `try_emplace` on the parent.

andrewtoth commented:
Rebased due to #34164.
Make `m_connect_block_view` be a `CoinsViewCacheNonMutating` instead of creating an ephemeral `CoinsViewCacheNonMutating`.

`git range-diff ab233255d444ccf6ffe4a45cb02bfc3e5fb71bdb..2f100b16b7bb09572ed3b04aa1484523af7509d9 6750744eb32da428b0a44c6eaab69e5741a25167..0e45e759e0197f01b82689298a164d442264fc8a`

l0rinc reviewed (approved):
ACK 77b1f5a3a3bba44651aa5da5926a37d6eca9e863

The rebase since my last ack reordered the code a bit to work with the reusable `m_connect_block_view`.
I have re-reviewed every commit, left small nits, LGTM!

sedited commented on src/test/coinsviewcachenonmutating_tests.cpp:
```diff
@@ -0,0 +1,161 @@
+    for (const auto& tx : block.vtx) {
+        for (const auto& in : tx->vin) {
+            const auto& c{view.AccessCoin(in.prevout)};
+            BOOST_CHECK(c.IsSpent());
```
Nit: How about checking that `HaveCoin` and `GetCoin` also return nothing here (and same with the other test below)?

sedited reviewed (commented):
Approach ACK

Not wasting cache space for a potentially invalid block seems like another nice improvement in of itself.

andrewtoth committed:
fuzz: add target for CoinsViewOverlay

Co-authored-by: l0rinc <pap.lorinc@gmail.com>

l0rinc commented on src/coinsviewoverlay.h:
```diff
@@ -0,0 +1,32 @@
+class CoinsViewOverlay : public CCoinsViewCache
+{
+private:
+    std::optional<Coin> GetCoinFromBase(const COutPoint& outpoint) const noexcept override
```
Isn't that the point of such a specifier that we explicitly promise something that would be hard to deduce otherwise?

sedited replied:
I would say so in some circumstances, but in a virtual class that has its member functions commonly overridden by other components that seems like an anti-pattern.

andrewtoth committed:
fuzz: move backend mutating block to end of coins_view

Refactor TestCoinsView() to move code that directly modifies
backend_coins_view to the end of the function.
This prepares for a CoinsViewOverlay fuzz target that asserts
the backend_coins_view is not mutated by any methods before
BatchWrite is called.

Co-authored-by: l0rinc <pap.lorinc@gmail.com>

andrewtoth commented:
Thank you for your review @sipa. I've taken your suggestions.

- Moved `CoinsViewOverlay` to `coins.h` instead of its own file.
- Use a `CCoinsViewCache` subclass as a middle layer in the `CoinsViewOverlay` fuzz target. The subclass `MutationGuardCoinsViewCache` asserts that nothing mutates the internal `cacheCoins` except calls to `Flush` and `Sync`. Had to add some try/catch blocks to handle `CCoinsViewCache` as the backing layer and move code that directly mutates the backing layer.

`git diff 2f9b52fa348242db475f4aff1fee951fc74e934e..2790e6e71f62eb0a7a5f239d1f8ea933be526d88`

l0rinc commented on src/test/fuzz/coins_view.cpp:
9df24c2a635087037100b7b93828ca34d3f979d2
> This is a slightly modified version of the first two commits of https://github.com/bitcoin/bitcoin/pull/31132,

This commit necessitates some PR description updates I think

l0rinc reviewed (changes requested):
Most of the change seems fine to me, but I don't like the new fuzzers as they are, we're testing more invalid states now.
Instead of catching `std::logic_error` we should investigate where our setup wasn't realistic and fix that instead of just yolo-mutating the state in ways that the production code cannot. The tests should help increase our confidence in the code, catching `std::logic_error` won't help with that.
Landing https://github.com/bitcoin/bitcoin/pull/34124 and https://github.com/bitcoin/bitcoin/pull/33018 could probably help a bit in that regard.

andrewtoth-exo commented on src/test/fuzz/coins_view.cpp:
```diff
@@ -194,8 +259,12 @@ void TestCoinsView(FuzzedDataProvider& fuzzed_data_provider, CCoinsViewCache& co
+            std::unique_ptr<CCoinsViewCursor> coins_view_cursor = backend_coins_view.Cursor();
+            assert(is_db == !!coins_view_cursor);
+        } catch (const std::logic_error&) {
+            assert(dynamic_cast<CCoinsViewCache*>(&backend_coins_view));
```
Yeah I'm not sure this check adds much to the test. Just gating it on `is_db` seems acceptable to me.

andrewtoth commented on src/test/fuzz/coins_view.cpp:
```diff
@@ -73,10 +124,24 @@ void TestCoinsView(FuzzedDataProvider& fuzzed_data_provider, CCoinsViewCache& co
+                try {
+                    coins_view_cache.Flush(/*reallocate_cache=*/fuzzed_data_provider.ConsumeBool());
+                } catch (const std::logic_error& e) {
+                    assert(e.what() == std::string{"FRESH flag misapplied to coin that exists in parent cache"});
```
I keep the error caught in the backend cache's `BatchWrite`, and don't rethrow it.

l0rinc reviewed (approved):
lightly tested + code review ACK 70e93a10b7fad61483bc9044b483460a7faaa1b8

I couldn't test the fuzz locally on a mac, but played around with the rest, LGTM.
I hope we can clean up the fuzzer in the future to have even fewer invalid states, but this is a lot better than before.

sipa reviewed (commented):
ACK 70e93a10b7fad61483bc9044b483460a7faaa1b8, with a small suggestion

sedited reviewed (approved):
ACK 70e93a10b7fad61483bc9044b483460a7faaa1b8

willcl-ark reviewed (approved):
tACK 70e93a10b7fad61483bc9044b483460a7faaa1b8

Had this branch reindex 50k blocks and then left it to run overnight for a few newly-seen blocks without issue. The code changes generally look good to me, and the rationale (as a building block of #31132) makes sense.

I did not run any of the fuzz tests.

andrewtoth commented on src/test/coins_tests.cpp:
```diff
@@ -1159,4 +1159,25 @@ BOOST_AUTO_TEST_CASE(ccoins_reset_guard)
     BOOST_CHECK(!root_cache.HaveCoinInCache(outpoint));
 }
 
+BOOST_AUTO_TEST_CASE(ccoins_peekcoin)
```
Thanks @sipa. I think to not invalidate the ACKs here I will leave this change to a follow-up, unless you insist it needs to go in this PR.

andrewtoth replied:
@l0rinc I applied the diff locally and fuzzed the target for half an hour (~1.7 million iterations). No issues :+1:

andrewtoth replied:
Added this now and added you as co-author.

ryanofsky reviewed (approved):
Code review ACK 70e93a10b7fad61483bc9044b483460a7faaa1b8. Behavior change seems very good and I did not see any problems with the implementation. But it does seem like it could be simplified. Like I don't understand why this wouldn't just add a bool use_peek option to `CCoinsViewCache` instead of introducing a new `CoinsViewOverlay` class and `GetCoinFromBase` override. Also I don't understand why `CoinsViewOverlay` seems to be tested in some different and more complicated ways than  `CCoinsViewCache`. (Details in comments below.)

I could be missing some context, and this PR does have a lot of acks so may not be worth changing. But I suspect various simplifications and cleanups could be possible here, and maybe done in followups.

l0rinc commented on src/test/fuzz/coins_view.cpp:
```diff
@@ -212,7 +212,7 @@ void TestCoinsView(FuzzedDataProvider& fuzzed_data_provider, CCoinsViewCache& co
                 bool expected_code_path = false;
                 try {
-                    auto cursor{CoinsViewCacheCursor(sentinel, coins_map, /*will_erase=*/true)};
+                    auto cursor{CoinsViewCacheCursor(sentinel, coins_map, /*will_erase=*/fuzzed_data_provider.ConsumeBool())};
```
It's related to https://github.com/bitcoin/bitcoin/pull/34165#discussion_r2765710685 - but I don't mind if we add it in a separate PR

andrewtoth replied:
Removed this for now. Agree it's unrelated and can be done in a follow-up.

andrewtoth committed:
coins: introduce CoinsViewOverlay

Introduce `CoinsViewOverlay`, a `CCoinsViewCache` subclass that reads
coins without mutating the underlying cache via `FetchCoin()`.

Use `PeekCoin()` to look up a Coin through a stack of `CCoinsViewCache` layers without populating parent caches. This prevents the main cache from caching inputs pulled from disk for a block that has not yet been fully validated. Once `Flush()` is called on the view, these inputs will be added as spent to `coinsCache` in the main cache via `BatchWrite()`.

This is the foundation for async input fetching, where worker threads must not
mutate shared state.

Co-authored-by: l0rinc <pap.lorinc@gmail.com>

andrewtoth committed:
coins: don't mutate main cache when connecting block

Use `CoinsViewOverlay` when connecting blocks in `ConnectTip`.

Add a new integration test to verify that using
CoinsViewOverlay does not mutate the main cache
during validation for an invalid block.

Co-authored-by: l0rinc <pap.lorinc@gmail.com>

andrewtoth commented:
Thank you for your review @ryanofsky! I have taken some of your suggestions and left clarifying comments for others.

`git diff 70e93a10b7fad61483bc9044b483460a7faaa1b8..6ca865fb2bd82b1034c6946d27640999f8135c89`

andrewtoth commented on src/test/fuzz/coins_view.cpp:
```diff
@@ -35,6 +38,54 @@ bool operator==(const Coin& a, const Coin& b)
+        try {
+            CCoinsViewCache::BatchWrite(cursor, block_hash);
+        } catch (const std::logic_error& e) {
+            // This error is thrown if we try and write a fresh entry when we already have the same fresh entry.
```
Clarified the comment here to better explain this case. I hope you find it to be an improvement.

ryanofsky reviewed (commented):
Code review 5b524f311ed74a751040494735b9a47f5c16f0b2. Thanks for the updates and explanations. Since last review, `GetCoin` was changed to delegate to `PeekCoin` instead of the reverse, which resolves my original concern about `PeekCoin` implementations unintentionally modifying the cache, but raises other concerns (see below). Other than that, some comments were just moved and added, some code moved between commits, and GetCoinFromBase was renamed (thanks!).

If my concerns about `GetCoin` calling `PeekCoin` aren't real or aren't important, would be happy to re-ack. Otherwise it might be better to go back to the previous approach, or just make these methods pure-virtual and accept a little boilerplate in the subclasses.

At a very high level, I will say the approach of making sure connect block code does not modify the cache by passing a different coinsview at runtime does not seem ideal to me. More ideal would be for the coins view to have a const method for looking up without modifying the cache, and a non-const method for looking up with caching, and the connect block code would be changed directly to use non-caching lookups, and the compiler would enforce this because it is passed a const view. Still, this runtime implementation seems like an improvement over status quo, and I'm probably not seeing the bigger picture here since these changes are used in follow up PRs

andrewtoth commented on src/coins.cpp:
```diff
@@ -5,16 +5,17 @@
 
-std::optional<Coin> CCoinsView::GetCoin(const COutPoint& outpoint) const { return std::nullopt; }
+std::optional<Coin> CCoinsView::GetCoin(const COutPoint& outpoint) const { return PeekCoin(outpoint); }
+std::optional<Coin> CCoinsView::PeekCoin(const COutPoint& outpoint) const { return std::nullopt; }
```
Indeed, I had this thought too after I changed it. I don't think either approach is really ideal on its own, and I appreciate #34124 more now after trying to solve this. In #34124 we can make only PeekCoin pure virtual, which will delegate to GetCoin. Or just make them non-virtual and add better hooks as you recommended. But, I suppose this is better discussed in that PR.

For now, I think I will revert to the previous change (which had 5 or 6 ACKs). There is no path I can see that calls a mutating GetCoin after PeekCoin, and this would have been revealed by fuzzing.

andrewtoth commented on src/txdb.cpp:
```diff
@@ -74,6 +74,11 @@ std::optional<Coin> CCoinsViewDB::GetCoin(const COutPoint& outpoint) const
 
+std::optional<Coin> CCoinsViewDB::PeekCoin(const COutPoint& outpoint) const
+{
+    return GetCoin(outpoint);
```
I think this is only safe if `PeekCoin` is a pure virtual method so subclasses can't miss overriding it.

l0rinc commented on src/test/fuzz/coins_view.cpp:
```diff
@@ -35,18 +38,73 @@ bool operator==(const Coin& a, const Coin& b)
+    struct CacheCoinSnapshot {
+        COutPoint outpoint;
+        bool dirty{false};
+        bool fresh{false};
```
what are these used for here?

l0rinc commented on src/coins.cpp:
```diff
@@ -393,3 +408,8 @@ bool CCoinsViewErrorCatcher::HaveCoin(const COutPoint& outpoint) const
     return ExecuteBackedWrapper<bool>([&]() { return CCoinsViewBacked::HaveCoin(outpoint); }, m_err_callbacks);
 }
+
+std::optional<Coin> CCoinsViewErrorCatcher::PeekCoin(const COutPoint& outpoint) const
```
This still stands

l0rinc commented on src/test/fuzz/coins_view.cpp:
```diff
@@ -220,8 +253,10 @@ void TestCoinsView(FuzzedDataProvider& fuzzed_data_provider, CCoinsView& backend
+        if (is_db) {
+            std::unique_ptr<CCoinsViewCursor> coins_view_cursor = backend_coins_view.Cursor();
+            assert(!!coins_view_cursor);
+        }
```
this is kinda broken currently ([the backend isn't actually changed](https://github.com/bitcoin/bitcoin/pull/34124#discussion_r2809051500)) - do we really need to modify this here or can we wait for #34124 to land first and have a more meaningful assertion here?

sipa reviewed (commented):
reACK cae6d895f8a8cf5f57e05519536fda5d62b10841

sedited reviewed (approved):
Re-ACK cae6d895f8a8cf5f57e05519536fda5d62b10841

willcl-ark reviewed (approved):
ACK cae6d895f8a8cf5f57e05519536fda5d62b10841

LGTM still. Changes since last review (`git range-diff 70e93a1...cae6d895f8a`) in the commit reorg are clearer method names (e.g. GetCoinFromBase renamed to FetchCoinFromBase), some commit comments moved into code comments, which seems more useful, and some fuzz tidy-ups.

vasild reviewed (approved):
Cursory ACK cae6d895f8a8cf5f57e05519536fda5d62b10841

I don't get it why the default `PeekCoin()` is calling `GetCoin()`.

ryanofsky reviewed (approved):
Code review ACK cae6d895f8a8cf5f57e05519536fda5d62b10841. PR is basically back to the form I had acked the first time, implementing `PeekCoin()` by calling `GetCoin()`. This is not ideal because `PeekCoin()` is not supposed to modify caches and `GetCoin()` does that, but it at least avoids problems of the subsequent approach tried where `GetCoin()` calls `PeekCoin` and would result in bugs when subclasses implement `GetCoin` forgetting to override `PeekCoin`. Hopefully #34124 can clean all of this by making relevant methods pure virtual.

More broadly, I don't understand a lot of complexity behind the design in this PR. I don't understand why a new virtual `FetchCoinFromBase` virtual method, a new `CoinsViewOverlay` subclass, and new test suite `coinsviewoverlay_tests.cpp` need to be be written just to pass the connect block code a view that calls `Peek` instead of `Get`. It seems like you could add a simple `bool call_base_peek_instead_of_get` option to the existing `CCoinsViewCache` class with a 3 line diff and have all the benefits of these changes. But I understand followups extend `CoinsViewOverlay` and probably provide more justification.

I also don't think swapping in a different `m_connect_block_view` class at runtime seems like the ideal way of preventing connect block code from modifying caches, when it should be possible to have const/nonconst methods or const/mutable views enforcing this at compile time. But I understand this would be a more invasive and potentially larger overall change and probably not a priority.

Overall looks like PR is ready for merge but would be good to respond to latest batch of comments from l0rinc https://github.com/bitcoin/bitcoin/pull/34165#issuecomment-3905465222. Thanks for all the followups on my previous reviews and answering all my questions!

optout21 commented:
Post-merge ACK, and documenting some concerns:

- Naming: `CoinsViewOverlay` misses the 'C' prefix used by most other classes in this domain. I find the 'Overlay' qualifier rather vague. Since it is used in only one place, naming based on the usage (as opoosed to features) seems appropriate to me here, sg. like "CCoinsCacheBlockView" (note: the variable holding it is named `m_connect_block_view`).
  (https://github.com/bitcoin/bitcoin/pull/31132/changes/3c7fcf10e97ba1c4d703bb4c3a04390d0e492367#r3187117587 #31132 )
- Since `CoinsViewOverlay` is effectively used in only one place, separating it into it's own .h/.cpp file would be an improvement, as less build units would be affected by it
  (https://github.com/bitcoin/bitcoin/pull/31132/changes/3c7fcf10e97ba1c4d703bb4c3a04390d0e492367#r3186363301 #31132 ).
